### ISSUE TITLE
[AIR] Tracking non-feature, non-label columns through model training and prediction

### ISSUE THREAD

**Issue by bdewilde** [ray-project/ray#33647]
### Description

I have a `Dataset` consisting of many feature columns, a label column, and a couple other columns that, when combined, form a unique identifier for each row. I also have a `Preprocessor` that featurizes many of the "raw" feature columns before passing them into an `sklearn` estimator. I'm looking for a straightforward, consistent way to track the non-feature, non-label columns through preprocessing, model training, and (batch) prediction alongside their corresponding features and labels. This is a fairly common scenario afaik, but I wasn't able to find any representative examples (e.g. [this example](https://docs.ray.io/en/latest/ray-air/examples/sklearn_example.html) is features+label only). A while back I submitted a related but separate issue about this [here](https://github.com/ray-project/ray/issues/30261).

Previously, I used an earlier version of `ray` (2.0ish) to implement a subclass of `SklearnTrainer` that split out and handled the identifier columns, similar to how it was already handling the label column, and also similar to how `BatchPredictor.predict()` had a `keep_columns` argument. This stopped working in v2.2 or 2.3 (not sure -- possibly [this PR](https://github.com/ray-project/ray/pull/31585)?), when the preprocessor started getting applied earlier in the process, which means that `keep_columns` is now drawn from the output of the preprocessed dataset rather than from the input dataset.

So. I'm wondering if I should update my preprocessor to pass through the identifier columns and also update my subclassed trainer to pull them out of the preprocessed data before fitting my estimator. But mostly I'm hoping for a simpler, cleaner implementation that doesn't require custom classes or user hassle. Is this functionality already implemented in `ray`, and I've just missed it?! If not... could it be? 🙏

**Comment by bdewilde** [comment]
Update: I've since spent a full day trying to work within the current framework, without success. I'd like to have a `Preprocessor` that works on both labeled and unlabeled data, as applied to the full `Dataset`s before splitting off any label and/or "keep" columns prior to model training/inference. One problem I've hit is that the labeled data includes, well, a label column, while the unlabeled data does not, and I don't see an easy way to accommodate that. (Looks like this has been submitted as [an issue](https://github.com/ray-project/ray/issues/28617) before.) I also don't see any way to "union" preprocessors applied to different sets of columns; one can chain or concatenate, but that's generally not equivalent. I guess I could try something like `Dataset.select_columns(keep_columns).zip()`? Maybe there are not-too-complicated solutions to these common situations, but I've not found them anywhere in the docs, and I don't see a way to implement them without custom subclasses for preprocessors and trainers. I'd love some help!

**Comment by amogkam** [comment]
Thanks @bdewilde this is great feedback!

As a workaround, you can do this to keep the original columns after preprocessing:

```python
preprocessor = checkpoint.get_preprocessor()

def manual_preprocessing(original_df):
   transformed_df = preprocessor.transform_batch(original_df)
   # Add back the original columns that you want to keep.
   transformed_df[original_columns] = original_df[original_columns]
    return transformed_df

transformed_ds = ds.map_batches(manual_preprocessing, batch_format="pandas")

# Don't apply preprocessing twice during prediction.
checkpoint.set_preprocessor(None) # Ray 2.4+
checkpoint.set_preprocessor(BatchMapper(lambda x: x)) # Ray <2.4

predictor = BatchPredictor.from_checkpoint(checkpoint)
predictor.predict(transformed_ds, feature_columns=[...], keep_columns=[...])
```

**Comment by bdewilde** [comment]
Hi @amogkam , just looping back around to say that this is a cute trick, and I did make it work for my use case! 🎉

That said, it would be great to not have to checkpoint-hack for this sort of situation. I've been doing ML for many years now, and I almost always have some unique identifier(s) alongside features and (for supervised tasks) target values in my training data that I want to confidently join up with predictions, to facilitate evaluation and downstream usage of the outputs. Any built-in solution would be much appreciated!

**Comment by amogkam** [comment]
Makes sense, we will make the change so that preprocessors doesn't override existing columns!
